File: Aladdin-ai--main/tool-system/src/main/kotlin/com/aladdin/tools/tools/DiscordTool.kt
package com.aladdin.tools.tools
import javax.inject.Singleton
import javax.inject.Inject
import dagger.hilt.android.qualifiers.ApplicationContext

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.aladdin.tools.tools.BaseTool
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.MediaType.Companion.toMediaType
import org.json.JSONObject

/**
 * Phase 9 — Discord Tool
 * Send messages, list servers/channels, fetch messages, react via Discord REST API.
 */
@Singleton
class DiscordTool @Inject constructor(@ApplicationContext private val context: Context) : BaseTool {

    override val id = "discord"

    override val name = "discord"
    override val description = "Send Discord messages, manage channels, read notifications, upload files, react"

    private val botToken get() = System.getenv("DISCORD_BOT_TOKEN") ?: ""
    private val apiBase = "https://discord.com/api/v10"
    private val httpClient = OkHttpClient()

    private suspend fun apiCall(method: String, path: String, payload: JSONObject? = null): JSONObject =
        withContext(Dispatchers.IO) {
            if (botToken.isEmpty()) throw RuntimeException("DISCORD_BOT_TOKEN not set")
            val builder = Request.Builder()
                .url("$apiBase$path")
                .header("Authorization", "Bot $botToken")
                .header("Content-Type", "application/json")
            when (method) {
                "POST" -> builder.post((payload ?: JSONObject()).toString()
                    .toRequestBody("application/json".toMediaType()))
                "PUT" -> builder.put((payload ?: JSONObject()).toString()
                    .toRequestBody("application/json".toMediaType()))
                "DELETE" -> builder.delete()
                else -> builder.get()
            }
            val resp = httpClient.newCall(builder.build()).execute()
            try { JSONObject(resp.body?.string() ?: "{}") } catch (e: Exception) { JSONObject() }
        }

    suspend fun sendMessage(channelId: String, content: String, replyToId: String? = null): ToolResult {
        return try {
            val payload = JSONObject().apply {
                put("content", content)
                replyToId?.let { put("message_reference", JSONObject().put("message_id", it)) }
            }
            val result = apiCall("POST", "/channels/$channelId/messages", payload)
            ToolResult.success(id, JSONObject().apply {
                put("message_id", result.optString("id"))
                put("channel_id", channelId); put("sent", true)
            }.toString())
        } catch (e: Exception) { ToolResult.error(id, "Discord send error: ${e.message}") }
    }

    suspend fun getMessages(channelId: String, limit: Int = 20): ToolResult {
        return try {
            val result = apiCall("GET", "/channels/$channelId/messages?limit=${minOf(limit, 100)}")
            ToolResult.success(id, JSONObject().apply {
                put("raw", result.toString()); put("channel_id", channelId)
            }.toString())
        } catch (e: Exception) { ToolResult.error(id, e.message ?: "Error") }
    }

    suspend fun addReaction(channelId: String, messageId: String, emoji: String): ToolResult {
        return try {
            val encoded = java.net.URLEncoder.encode(emoji, "UTF-8")
            apiCall("PUT", "/channels/$channelId/messages/$messageId/reactions/$encoded/@me")
            ToolResult.success(id, JSONObject().apply {
                put("reacted", emoji); put("message_id", messageId)
            }.toString())
        } catch (e: Exception) { ToolResult.error(id, e.message ?: "Reaction error") }
    }

    suspend fun listServers(): ToolResult {
        return try {
            val result = apiCall("GET", "/users/@me/guilds")
            ToolResult.success(id, JSONObject().put("servers", result.toString()).toString())
        } catch (e: Exception) { ToolResult.error(id, e.message ?: "Error") }
    }

    fun openDiscord(): ToolResult {
        return try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse("discord://")).apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            }
            context.startActivity(intent)
            ToolResult.success(id, JSONObject().put("discord_opened", true).toString())
        } catch (e: Exception) { ToolResult.error(id, e.message ?: "Open failed") }
    }

    override suspend fun execute(params: Map<String, String>): ToolResult {
        return when (val action = (params["action"] ?: "send")) {
            "send" -> sendMessage((params["channel_id"] ?: return ToolResult.error(id, "Missing required parameter: " + "channel_id")), (params["content"] ?: return ToolResult.error(id, "Missing required parameter: " + "content")),
                (params["reply_to"] ?: "").ifEmpty { null })
            "messages" -> getMessages((params["channel_id"] ?: return ToolResult.error(id, "Missing required parameter: " + "channel_id")), (params["limit"]?.toIntOrNull() ?: 20))
            "react" -> addReaction((params["channel_id"] ?: return ToolResult.error(id, "Missing required parameter: " + "channel_id")),
                (params["message_id"] ?: return ToolResult.error(id, "Missing required parameter: " + "message_id")), (params["emoji"] ?: return ToolResult.error(id, "Missing required parameter: " + "emoji")))
            "servers" -> listServers()
            "open" -> openDiscord()
            else -> ToolResult.error(id, "Unknown Discord action: $action")
        }
    }
}

File: Aladdin-ai--main/backend/tools/discord_tool.py
"""Discord tool — send messages, manage servers/channels, voice, reactions via discord.py."""
from __future__ import annotations
import asyncio, logging, os, time
from pathlib import Path
from typing import Dict, List, Optional
from .base import BaseTool, ToolResult

log = logging.getLogger(__name__)


def _get_token() -> str:
    token = os.getenv("DISCORD_BOT_TOKEN", "")
    if not token:
        raise RuntimeError("DISCORD_BOT_TOKEN not set")
    return token


async def _http(method: str, endpoint: str, **kwargs):
    """Thin async wrapper around the Discord REST API."""
    import aiohttp
    token = _get_token()
    base = "https://discord.com/api/v10"
    headers = {"Authorization": f"Bot {token}", "Content-Type": "application/json"}
    async with aiohttp.ClientSession(headers=headers) as session:
        fn = getattr(session, method)
        async with fn(f"{base}{endpoint}", **kwargs) as resp:
            if resp.status in (200, 201):
                return await resp.json()
            text = await resp.text()
            raise RuntimeError(f"Discord API error {resp.status}: {text}")


class SendDiscordMessageTool(BaseTool):
    name = "send_discord_message"
    description = "Send a text message to a Discord channel."
    parameters = {"type": "object", "properties": {
        "channel_id": {"type": "string"},
        "content": {"type": "string"},
        "reply_to_message_id": {"type": "string", "description": "Optional message ID to reply to"}},
        "required": ["channel_id", "content"]}

    async def execute(self, channel_id: str, content: str, reply_to_message_id: str = None) -> ToolResult:
        t0 = time.time()
        try:
            payload: Dict = {"content": content}
            if reply_to_message_id:
                payload["message_reference"] = {"message_id": reply_to_message_id}
            data = await _http("post", f"/channels/{channel_id}/messages", json=payload)
            return ToolResult(True, self.name, {"message_id": data["id"], "channel_id": channel_id},
                              duration_ms=(time.time() - t0) * 1000)
        except Exception as exc:
            return ToolResult(False, self.name, error=str(exc), duration_ms=(time.time() - t0) * 1000)


class SendDiscordFileTool(BaseTool):
    name = "send_discord_file"
    description = "Upload and send a file to a Discord channel."
    parameters = {"type": "object", "properties": {
        "channel_id": {"type": "string"}, "file_path": {"type": "string"},
        "message": {"type": "string", "default": ""}},
        "required": ["channel_id", "file_path"]}

    async def execute(self, channel_id: str, file_path: str, message: str = "") -> ToolResult:
        t0 = time.time()
        try:
            import aiohttp
            token = _get_token()
            headers = {"Authorization": f"Bot {token}"}
            p = Path(file_path)
            with open(p, "rb") as f:
                form = aiohttp.FormData()
                if message:
                    form.add_field("content", message)
                form.add_field("file", f, filename=p.name)
                async with aiohttp.ClientSession() as session:
                    async with session.post(
                        f"https://discord.com/api/v10/channels/{channel_id}/messages",
                        headers=headers, data=form
                    ) as resp:
                        data = await resp.json()
            return ToolResult(True, self.name, {"message_id": data.get("id"), "file": p.name},
                              duration_ms=(time.time() - t0) * 1000)
        except Exception as exc:
            return ToolResult(False, self.name, error=str(exc), duration_ms=(time.time() - t0) * 1000)


class GetDiscordMessagesTool(BaseTool):
    name = "get_discord_messages"
    description = "Fetch recent messages from a Discord channel."
    parameters = {"type": "object", "properties": {
        "channel_id": {"type": "string"}, "limit": {"type": "integer", "default": 20}},
        "required": ["channel_id"]}

    async def execute(self, channel_id: str, limit: int = 20) -> ToolResult:
        t0 = time.time()
        try:
            data = await _http("get", f"/channels/{channel_id}/messages", params={"limit": min(limit, 100)})
            messages = [
                {"id": m["id"], "author": m["author"]["username"],
                 "content": m["content"], "timestamp": m["timestamp"]}
                for m in data
            ]
            return ToolResult(True, self.name, {"messages": messages, "count": len(messages)},
                              duration_ms=(time.time() - t0) * 1000)
        except Exception as exc:
            return ToolResult(False, self.name, error=str(exc), duration_ms=(time.time() - t0) * 1000)


class AddDiscordReactionTool(BaseTool):
    name = "add_discord_reaction"
    description = "React to a Discord message with an emoji."
    parameters = {"type": "object", "properties": {
        "channel_id": {"type": "string"}, "message_id": {"type": "string"},
        "emoji": {"type": "string", "description": "Emoji character or name:id"}},
        "required": ["channel_id", "message_id", "emoji"]}

    async def execute(self, channel_id: str, message_id: str, emoji: str) -> ToolResult:
        t0 = time.time()
        try:
            import urllib.parse
            encoded = urllib.parse.quote(emoji)
            await _http("put", f"/channels/{channel_id}/messages/{message_id}/reactions/{encoded}/@me")
            return ToolResult(True, self.name, {"reacted": emoji, "message_id": message_id},
                              duration_ms=(time.time() - t0) * 1000)
        except Exception as exc:
            return ToolResult(False, self.name, error=str(exc), duration_ms=(time.time() - t0) * 1000)


class ListDiscordServersTool(BaseTool):
    name = "list_discord_servers"
    description = "List all Discord servers (guilds) the bot is a member of."
    parameters = {"type": "object", "properties": {}}

    async def execute(self) -> ToolResult:
        t0 = time.time()
        try:
            guilds = await _http("get", "/users/@me/guilds")
            servers = [{"id": g["id"], "name": g["name"], "icon": g.get("icon")} for g in guilds]
            return ToolResult(True, self.name, {"servers": servers, "count": len(servers)},
                              duration_ms=(time.time() - t0) * 1000)
        except Exception as exc:
            return ToolResult(False, self.name, error=str(exc), duration_ms=(time.time() - t0) * 1000)


class ListDiscordChannelsTool(BaseTool):
    name = "list_discord_channels"
    description = "List all channels in a Discord server."
    parameters = {"type": "object", "properties": {
        "guild_id": {"type": "string"}}, "required": ["guild_id"]}

    async def execute(self, guild_id: str) -> ToolResult:
        t0 = time.time()
        try:
            channels = await _http("get", f"/guilds/{guild_id}/channels")
            result = [{"id": c["id"], "name": c["name"], "type": c["type"]} for c in channels]
            return ToolResult(True, self.name, {"channels": result, "count": len(result)},
                              duration_ms=(time.time() - t0) * 1000)
        except Exception as exc:
            return ToolResult(False, self.name, error=str(exc), duration_ms=(time.time() - t0) * 1000)
